lib/uplinkc/bucket.go

```diff
@@ -0,0 +1,193 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+// #include "uplink_definitions.h"
+import "C"
+
+import (
+	"unsafe"
+
+	"storj.io/storj/lib/uplink"
+	"storj.io/storj/pkg/storj"
+)
+
+// Bucket is a scoped uplink.Bucket
+type Bucket struct {
+	scope
+	*uplink.Bucket
+}
+
+// create_bucket creates a new bucket if authorized.
+//export create_bucket
+func create_bucket(projectHandle C.ProjectRef_t, name *C.char, bucketConfig *C.BucketConfig_t, cerr **C.char) C.BucketInfo_t {
+	project, ok := universe.Get(projectHandle._handle).(*Project)
+	if !ok {
+		*cerr = C.CString("invalid project")
+		return C.BucketInfo_t{}
+	}
+
+	var config *uplink.BucketConfig
+	if bucketConfig != nil {
+		config = &uplink.BucketConfig{
+			PathCipher: storj.CipherSuite(bucketConfig.path_cipher),
+			EncryptionParameters: storj.EncryptionParameters{
+				CipherSuite: storj.CipherSuite(bucketConfig.encryption_parameters.cipher_suite),
+				BlockSize:   int32(bucketConfig.encryption_parameters.block_size),
+			},
+		}
+		config.Volatile.RedundancyScheme = storj.RedundancyScheme{
+			Algorithm:      storj.RedundancyAlgorithm(bucketConfig.redundancy_scheme.algorithm),
+			ShareSize:      int32(bucketConfig.redundancy_scheme.share_size),
```

Comment:
I'm not sure how it works with cgo but will `bucketConfig.redundancy_scheme` be always allocated? or it can be a nil?

Reply:
Here it cannot be nil.

Reply:
So, I've not yet conclusively decided. We can do it two ways:
1. keep a pointer, this means that from C it will be more annoying to use - you need to alloc/free the config
2. no pointer, this means in Go we need to detect `0` case and handle that instead of missing RS.

Reply:
We didn't make it a pointer because the libuplink code used a value. It might also be worth noting that this is in the volatile field which is subject to change and we weren't supposed to really mess with anyways.